docs/codeql/codeql-language-guides/customizing-library-models-for-python.rst

```diff
@@ -0,0 +1,449 @@
+
+Continued example: Multiple ways to obtain a type
+-------------------------------------------------
+
+The invoke package provides multiple ways to obtain a **Context** instance. The following example shows how to add a new way to obtain a **Context** instance:
+
+.. code-block:: python
+
+  from invoke import context
+  c = context.Context()
+  c.run(cmd) # <-- add 'cmd' as a taint sink
+
+Comparing to the previous Python snippet, the **Context** class is now found as **invoke.context.Context** instead of **invoke.Context**.
+We could add a data extension similar to the previous one, but with the type **invoke.context.Context**. However, we can also use the **typeModel** extensible predicate to describe how to reach **invoke.Context** from **invoke.context.Context**:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: typeModel
+      data:
+        - ["invoke.Context", "invoke", "Member[context].Member[Context].Instance"]
+
+- The first column, **"invoke.Context"**, is the name of the type to reach.
+- The second column, **"invoke"**, is the name of the type from which to evaluate the path.
+- The third column, **"Member[context].Member[Context].Instance"**, is the access path leading from **invoke** to **invoke.Context**.
+
+Combining this with the sink model we added earlier, the sink in the example is detected by the model.
+
+Example: Taint sources from Django 'upload_to' argument
+-------------------------------------------------------
+
+This example is a bit more advanced, involving both a callback function and a class constructor.
+The Django web framework allows you to specify a function that determines the path where uploaded files are stored (see the `Django documentation <https://docs.djangoproject.com/en/5.0/ref/models/fields/#django.db.models.FileField.upload_to>`_).
+This function is passed as an argument to the **FileField** constructor.
+The function is called with two arguments: the instance of the model and the filename of the uploaded file.
+This filename is what we want to mark as a taint source. An example use looks as follows:
+
+.. code-block:: python
+
+  from django.db import models
+
+  def user_directory_path(instance, filename): # <-- add 'filename' as a taint source
+    # file will be uploaded to MEDIA_ROOT/user_<id>/<filename>
+    return "user_{0}/{1}".format(instance.user.id, filename)
+
+  class MyModel(models.Model):
+    upload = models.FileField(upload_to=user_directory_path) # <-- the 'upload_to' parameter defines our custom function
+
+Note that this source is already known by the CodeQL Python analysis, but for this example, you could use the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/python-all
+        extensible: sourceModel
+      data:
+        - [
+            "django.db.models.FileField!",
+            "Call.Argument[upload_to:].Parameter[1,filenam:]",
```

Comment:
```suggestion
            "Call.Argument[0,upload_to:].Parameter[1,filenam:]",
```
It seems to me that you can also give the `upload_to` argument positionally.

Reply:
You are right. I got confused by the default argument (seen [here](https://docs.djangoproject.com/en/5.0/ref/models/fields/#filefield)).